doc/cprover-manual/contracts-invariant.md

```diff
@@ -0,0 +1,123 @@
+# Invariant Clause
+
+An _invariant_ clause specifies a property that must be preserved
+by every iteration of a loop.
+
+### Syntax
+
+An _invariant_ clause accepts a valid Boolean expression
+over the variables visible at the same scope as the loop.
+Additionally, [history variables] are also allowed within invariant clauses.
+
+```c
+__CPROVER_loop_invariant(bool cond)
+```
+
+Invariant clauses may be specified just after the loop guard.
+Multiple invariant clauses on the same loop is allowed,
+and is equivalent to a single invariant clause that is a conjunction of those clauses.
+A few examples are shown below.
+
+```c
+while(i < n)
+__CPROVER_loop_invariant(0 <= i && i <= n)
+{ ... }
+```
+
+```c
+for(int i = 0; i < n; ++i)
+__CPROVER_loop_invariant(0 <= i)
+__CPROVER_loop_invariant(i <= n)
+{ ... }
+```
+
+```c
+do { ... }
+while (i < n)
+__CPROVER_loop_invariant(0 <= i)
+__CPROVER_loop_invariant(i <= n);
+```
+
+**Important.** Invariant clauses must be free of _side effects_,
+for example, mutation of local or global variables.
+
+
+### Semantics
+
+The loop invariant clause expands to several assumptions and assertions:
+1. The invariant is asserted just before the first iteration
+2. The invariant is assumed on a non-deterministic state to model a non-determistic iteration
+3. The invariant is finally asserted again to establish its inductiveness
+
+To illustrate the key idea,
+consider the following [binary search] loop with invariant clauses:
+
+```c
+#include <assert.h>
+
+int bin_search(int val, int buf[], int size)
+{
+  if(size < 0) return -1;
+
+  long lb = 0, ub = size - 1;
+  long mid = (lb + ub) / 2;
+
+  while(lb <= ub)
+  __CPROVER_loop_invariant(0 <= lb && lb - 1 <= ub && ub < size)
+  __CPROVER_loop_invariant(mid == (lb + ub) / 2)
```

Comment:
This invariant seems counter-intuitive (this may be something much deeper than the actual cbmc checking though). There are points inside the loop where this invariant does not hold. For example, on completion of line 71 or 73  this invariant will typically not hold (although it will again after line 74 is completed).

I see this is shown below (how this is implemented), maybe a note that `__CPROVER_loop_invariant` is checked at each loop iteration and may not hold for the body of the loop?

Reply:
Thanks; I would add a not at the top explaining a bit more about how loop invariants work and why they help us obtain an unbounded correctness proof.

But, this is the standard interpretation of [loop invariants in Floyd-Hoare logic](https://en.wikipedia.org/wiki/Loop_invariant). Much like function contracts, loop contracts are enforced at boundary of the loop and may not hold at intermediate steps in the loop body. It must be preserved before the loop, and after each iteration of the loop, thereby providing an inductive argument that the invariant must _always_ hold after the loop, which is the key to proving any subsequent assertions.

Reply:
Added a note (lines 58 -- 64).